Fix: Improve continuous stream stability and handle mbedtls exceptions in ESP-IDF v5

tobiasguyer opened the pull request:
Context: With the recent update to ESP-IDF v5, there have been issues with continuous stream handling and mbedtls errors, leading to occasional connection failures, missed reads, missreads, and delays in reconnecting, wich lead to aborts after some tracks.
What was changed:

    Exception Handling:
        Added specific exception handling around mbedtls operations to catch and log any errors that occur during streaming.
        Known Issues: Currently, mbedtls might throw exceptions occasionally, but these are handled with minimal impact on stream stability. We are hoping for an upstream fix from Espressif to resolve these issues long-term.

    Reconnection Logic:
        Improved reconnection logic after an error occurs, reducing the time it takes to attempt a reconnect.
        Reconnection logic now includes exponential backoff to avoid overwhelming the system with retry attempts, providing smoother recovery after errors.

    Buffering:
        HTTPClient tries to recover from garbage missreads and keeps from parsing the buffer if no "HTTP/" header was found

    Improved Logging:
        Introduced logging around critical mbedtls errors to help debug the root cause of connection drops and other stream-related issues.

Known Issues:

    The stream is not yet perfect, as reconnection can still take longer than expected in some cases, leading to short audible gaps.

tobiasguyer committed:
fix: Update CivetWeb to v1.17.0 and migrate to pthreads

- Ported CivetWeb to version 1.17.0 to bring in latest features and fixes.
- Replaced FreeRTOS threading with POSIX pthreads
Notes:
- Tested on ESP32, WIN32.

tobiasguyer committed:
fix: adjust handleClose to avoid const issues in ESP-IDF v5

- Modified handleClose to maintain compatibility with ESP-IDF 5, addressing const-related issues with mg_connection.

tobiasguyer committed:
feat: Update getMacAddress for WIN32/64 compatibility

- Modified getMacAddress function to support WIN32/64 platforms.
- Ensures unique device ID generation across Windows environments.
- Platform-specific adjustments were made to retrieve MAC address on Windows.

tobiasguyer committed:
feat: Update WrappedSemaphore to support custom initial countValue

- Modified WrappedSemaphore constructor to accept an optional countValue parameter.
- Default value of countValue is still 0, but now can be initialized with a custom value.
- This change improves flexibility for different semaphore use cases.

tobiasguyer committed:
feat: Guard BELL_LOG with semaphore for thread safety

- Introduced a semaphore to protect BELL_LOG and ensure thread-safe logging.
- This change prevents race conditions when multiple threads access the log.
- The semaphore is used to synchronize log access and avoid concurrent writes.

tobiasguyer committed:
fix: Enhance error handling for mbedtls in ESP32

- Addressed issues causing garbage data in buffers due to mbedtls mismanagement.
- Improved handling of misread and miswrite operations, ensuring data integrity.
- Fixed error handling for misconnection scenarios, reducing failed or corrupted connections.
- Enhanced stability and reliability of secure communications on ESP32 by refining error reporting and flow control.

tobiasguyer committed:
Fix: changed color in BELL_LOG to string argument

tobiasguyer commented:
I've forgot to mention, that i've added a "fix" to espressif Framework. So if you've still got some issue, that would probably be a start. I'd love to link you to the Github comment, but i cant't find it no more. 
Espressif\frameworks\esp-idf-v5.3.1\components\mbedtls\mbedtls\library\ssl_msg.c: ~5981
```
static int ssl_write_real(mbedtls_ssl_context *ssl,

                          const unsigned char *buf, size_t len)
{
    int ret = mbedtls_ssl_get_max_out_record_payload(ssl);
    const size_t max_len = (size_t) ret;

    if (ret < 0) {
        MBEDTLS_SSL_DEBUG_RET(1, "mbedtls_ssl_get_max_out_record_payload", ret);
        return ret;
    }

    if (len > max_len) {
#if defined(MBEDTLS_SSL_PROTO_DTLS)
        if (ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM) {
            MBEDTLS_SSL_DEBUG_MSG(1, ("fragment larger than the (negotiated) "
                                      "maximum fragment length: %" MBEDTLS_PRINTF_SIZET
                                      " > %" MBEDTLS_PRINTF_SIZET,
                                      len, max_len));
            return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
        } else
#endif
        len = max_len;
    }

    if (ssl->out_left != 0) {
        /*
         * The user has previously tried to send the data and
         * MBEDTLS_ERR_SSL_WANT_WRITE or the message was only partially
         * written. In this case, we expect the high-level write function
         * (e.g. mbedtls_ssl_write()) to be called with the same parameters
         */
        if ((ret = mbedtls_ssl_flush_output(ssl)) != 0) {
            MBEDTLS_SSL_DEBUG_RET(1, "mbedtls_ssl_flush_output", ret);
            return ret;
        }
        return 0; // ======================> here should be this !!!!
    } else {
        /*
         * The user is trying to send a message the first time, so we need to
         * copy the data into the internal buffers and setup the data structure
         * to keep track of partial writes
         */
        ssl->out_msglen  = len;
        ssl->out_msgtype = MBEDTLS_SSL_MSG_APPLICATION_DATA;
        if (len > 0) {
            memcpy(ssl->out_msg, buf, len);
        }

        if ((ret = mbedtls_ssl_write_record(ssl, SSL_FORCE_FLUSH)) != 0) {
            MBEDTLS_SSL_DEBUG_RET(1, "mbedtls_ssl_write_record", ret);
            return ret;
        }
    }

    return (int) len;
}

tobiasguyer committed:
fix: Add out-of-bounds check for header name and value in readResponseHeaders()

- Implemented bounds checking for the header name and value pointers in the HTTP response headers.
- Prevents potential out-of-bounds access when parsing response headers.
- Throws an std::out_of_range exception with descriptive message if header name or value exceeds the available buffer size.

This ensures the stability and correctness of the HTTP response parsing and handles edge cases where the headers might exceed the buffer size.

tobiasguyer committed:
feat: Reduced verbosity in error logs for cleaner output

tobiasguyer commented:
I can't say with absolute certainty that the memory leak is completely resolved, but I’ve observed no memory leaks during 18 hours of continuous playback.  

### **Context**
The primary issue with ESP-IDF v5 seemed to be that TCP PCB (Protocol Control Block) resources were not being freed when too many connections were running simultaneously. This behavior differs from ESP-IDF v4, where such resources were properly released. For reference, see the discussion in [[ESP32 Forum - TCP PCB Issue](https://esp32.com/viewtopic.php?t=31858)](https://esp32.com/viewtopic.php?t=31858).  

### **Solution**
The following changes in TCPSocket.h, combined with enabling `CONFIG_LWIP_SO_LINGER` in the menuconfig (`Component config -> LWIP -> Enable SO_LINGER processing`), appear to have mitigated the memory leak:  

```cpp
#if defined(ESP_PLATFORM)
// https://esp32.com/viewtopic.php?t=31858
struct linger linger_opt = {
    .l_onoff = 1,  // Enable linger: Must be 1 for l_linger time to be used.
    .l_linger = 0,
};
if (setsockopt(sockFd, SOL_SOCKET, SO_LINGER, &linger_opt,
               sizeof(linger_opt)) != 0) {
    close();
    BELL_LOG(error, "http", "Could not connect to %s. Error %d", host.c_str(),
             errno);
    throw std::runtime_error("Resolve failed");
}
#endif
```

### **References**
- For details on the SO_LINGER behavior and its potential role in fixing this issue, refer to [[ESP32 Forum - SO_LINGER Discussion](https://esp32.com/viewtopic.php?t=34920)](https://esp32.com/viewtopic.php?t=34920).

feelfreelinux commented:
Good find :) I'll try to take a closer look at your changes in the coming weeks, and consolidate it with an ongoing cleanup of bell and cspot code that I'm working on right now. I really appreciate your effort :)

tobiasguyer committed:
feat(civetweb): minimal thread-names for esp-threads

tobiasguyer committed:
feat(sink): add VS1053 sink to bell/main/audio-sinks

tobiasguyer committed:
fix(logger): guard BellLogger with mutex to avoid race conditions

Protect shared logging state with a mutex to avoid concurrent writes from multiple tasks.

tobiasguyer committed:
feat(http-server): add options vector to BellHTTPServer

- allow passing options as std::vector<std::pair<std::string, std::string>>
- keep API backward compatible with existing usage

tobiasguyer committed:
feat(http-client): extend HTTPClient API (and fix some minor things)

- add HTTPClient::Response::read() and readExact() to consume prefetched header bytes
- add optional cookieChar to HTTPClient Request and Response

tobiasguyer committed:
feat(nanopb-helper): extend NanoPBHelper API

- add helpers to convert to and from pbArray_t*
- add PB_MOVE_ASSIGN macro for move-style assignment

tobiasguyer committed:
fix(task): remove extra ESP32 pthread priority offset

Remove '+ CONFIG_ESP32_PTHREAD_TASK_PRIO_DEFAULT' from BellTask priority
to avoid overshooting expected task priority on ESP32.

tobiasguyer committed:
fix: add steps to github-workflow

tobiasguyer committed:
fix(nanopb): use message_factory instead of reflection.MakeClass
